packages/astro/src/manifest/virtual-module.ts

```diff
@@ -0,0 +1,111 @@
+import type { Plugin } from 'vite';
+import { CantUseManifestModule, ForbiddenManifestModule } from '../core/errors/errors-data.js';
+import { AstroError } from '../core/errors/index.js';
+import type { Logger } from '../core/logger/core.js';
+import type { AstroSettings } from '../types/astro.js';
+import type {
+	AstroConfig,
+	ClientConfigSerialized,
+	ServerConfigSerialized,
+} from '../types/public/index.js';
+
+const VIRTUAL_MODULES_IDS = {
+	client: 'astro:manifest/client',
+	server: 'astro:manifest/server',
+};
+
+const VIRTUAL_MODULES_IDS_SET = new Set(Object.values(VIRTUAL_MODULES_IDS));
+
+function resolveVirtualModuleId<T extends string>(id: T): `\0${T}` {
+	return `\0${id}`;
+}
```

Comment:
I think we could go with a simpler approach
```suggestion
const VIRTUAL_SERVER_ID = 'astro:manifest/server'
const RESOLVED_VIRTUAL_SERVER_ID = '\0' + VIRTUAL_SERVER_ID
const VIRTUAL_CLIENT_ID = 'astro:manifest/client'
const RESOLVED_VIRTUAL_CLIENT_ID = '\0' + VIRTUAL_CLIENT_ID
```